File: android/app/src/main/java/sol/vortexx/android/chat/api/MessageSender.kt
package sol.vortexx.android.chat.api

/**
 * Send path. Takes plaintext + room, returns a send outcome once the
 * frame has been queued on the WS (not when the server ACKs — delivery
 * receipts are a separate concern).
 */
interface MessageSender {
    suspend fun send(roomId: Long, plaintext: String): SendOutcome
}

sealed interface SendOutcome {
    data class Queued(val localId: Long) : SendOutcome
    data class Error(val reason: String) : SendOutcome
}

File: ios/Modules/Sources/Chat/api/MessageSender.swift
import Foundation
import DB

/// Send path. Returns a send outcome once the frame has been queued on
/// the WS — delivery receipts are a separate concern (Wave 13+).
public protocol MessageSender: Sendable {
    func send(roomId: Int64, plaintext: String) async -> SendOutcome
}

public enum SendOutcome: Sendable, Equatable {
    case queued(localId: Int64)
    case error(String)
}

/// Receive path — server-pushed message events, already decrypted and
/// persisted locally.
public protocol IncomingMessages: Sendable {
    func messagesIn(_ roomId: Int64) -> AsyncStream<[MessageRecord]>
}
